cellmlmanip/units.py

```diff
@@ -0,0 +1,451 @@
+"""Unit handling for CellML models, using the Pint unit library (replaces previous
+Sympy-units implementation
+"""
+import logging
+import math
+import numbers
+import os
+from functools import reduce
+from operator import mul
+from typing import Dict, List
+
+import pint
+import sympy
+from pint.converters import ScaleConverter
+from pint.definitions import UnitDefinition
+from sympy.printing.lambdarepr import LambdaPrinter
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+# The full list of supported CellML units
+# Taken from https://www.cellml.org/specifications/cellml_1.1/#sec_units
+CELLML_UNITS = {
+    # Base SI units
+    'ampere', 'candela', 'kelvin', 'kilogram', 'meter', 'mole', 'second',
+
+    # Derived SI units
+    'becquerel', 'celsius', 'coulomb', 'farad', 'gray', 'henry', 'hertz', 'joule',
+    'lumen', 'lux', 'newton', 'ohm', 'pascal', 'radian', 'siemens', 'sievert',
+    'steradian', 'tesla', 'volt', 'watt', 'weber',
+
+    'katal',
+
+    # Convenience units
+    'dimensionless', 'gram', 'liter',
+
+    # Aliases
+    'metre', 'litre',
+}
+
+CELLML_UNIT_PREFIXES = {
+    'yocto': 1e-24,
+    'zepto': 1e-21,
+    'atto':  1e-18,
+    'femto': 1e-15,
+    'pico':  1e-12,
+    'nano':  1e-9,
+    'micro': 1e-6,
+    'milli': 1e-3,
+    'centi': 1e-2,
+    'deci':  1e-1,
+    'deca':  1e+1,
```

Comment:
@jonc125 Are you happy to have this in as a silent extra option? I think we'll be allowing deca and deka in CellML 2.0 (also nobody uses it)